library/general/annotationrestriction/src.rego

```diff
@@ -0,0 +1,8 @@
+package requiredannotations
+
+violation[{"msg": msg, "details": {}}] {
+    #Retrieve the reference to the annotation you are looking for,
+    #compare it to the value you specified, denies if they do not match.
+    not input.review.object.metadata.annotations[input.parameters.annotationKey] == input.parameters.annotationValue
+    msg := sprintf("Create / update of <%v> blocked - Annotation key <%v> must have value '%v'.", [input.review.object.kind, input.parameters.annotationKey, input.parameters.annotationValue])
```

Comment:
We should avoid assuming the operation that was prevented ("create / update" here) or specifically naming the resource blocked.

The reason for this is there may be other contexts in which constraints are run (think CI/CD pipeline where create/update may not make sense.

Regarding the identity of the resource blocked, that should be up to the system to provide appropriate context as it may be implicit based on the context of the request (e.g. kubebuil create vs audit results).

An example of an external framework reporting identity is our audit results listing Group/Kind/Namespace/Name.

Reply:
@maxsmythe thanks for feedback. Kaleb and I are pretty new to OPA and Gatekeeper and we're having a hard time interpreting the feedback.

Are you asking for a change to drop the first part of the string like this?

```diff
-    msg := sprintf("Create / update of <%v> blocked - Annotation key <%v> must have value '%v'.", [input.review.object.kind, input.parameters.annotationKey, input.parameters.annotationValue])
+    msg := sprintf("Annotation key <%v> must have value '%v'.", [input.parameters.annotationKey, input.parameters.annotationValue])
```

I think the above change is a bad idea, as a user pushing a bundle of a deployment, service, and ingress into kubernetes with helm / kubectl / whatever, I'd like to know what is getting blocked. With the above change, as a user I wouldnt know if my deployment, service, or ingress is violating the policy. Is there a different way to get this information to the end user that we're missing?

Reply:
Ya, whatever's executing the code should do it for you:

Here it is clear that the namespace "hello" is being rejected b/c `kubectl create ns` is only trying to create one resource:

```
$ kubectl create ns hello
Error from server ([denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}): admission webhook "validation.gatekeeper.sh" denied the request: [denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}
```

Here we know which file is getting rejected because it's part of the Kubectl client's rejection message:
```
$ tree ns
ns
├── ns2.yaml
└── ns.yaml

0 directories, 2 files

$ kubectl apply -f ns
Error from server ([denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}): error when creating "ns/ns.yaml": admission webhook "validation.gatekeeper.sh" denied the request: [denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}
Error from server ([denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}): error when creating "ns/ns2.yaml": admission webhook "validation.gatekeeper.sh" denied the request: [denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}
```

Though reporting errors by file is a problem if you have a flat-file (e.g. `deploy/gatekeeper.yaml`):
```
$ cat ns.yaml
apiVersion: v1
kind: Namespace
metadata:
  name: namespace1
---

apiVersion: v1
kind: Namespace
metadata:
  name: namespace2

$ kubectl apply -f ns.yaml
Error from server ([denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}): error when creating "ns.yaml": admission webhook "validation.gatekeeper.sh" denied the request: [denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}
Error from server ([denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}): error when creating "ns.yaml": admission webhook "validation.gatekeeper.sh" denied the request: [denied by ns-must-have-gk] you must provide labels: {"gatekeeper"}
```

This should either be fixed by Gatekeeper specifying the failing resource, or changing the output of the kubectl client.

Then there is this line in the audit results:

```
  - enforcementAction: deny
    kind: Namespace
    message: 'you must provide labels: {"gatekeeper"}'
    name: kube-system
```
Where the violating resource is of kind "Namespace" and name "kube-system"

Ultimately the consideration is that it should be on the system not the constraint template author to specify which resource fails because:

1. Constraints could be run in multiple different contexts, and it's the context that knows what extra info is needed, not the constraint
2. Relying on template authors to provide this extra info will never be 100% effective as some authors will forget. A change to the framework can change the output of every constraint at once.

Reply:
> This should either be fixed by Gatekeeper specifying the failing resource, or changing the output of the kubectl client.

Is there already an issue to add the denied resource to the builtin gatekeeper validationwebhook message? I like that approach.

Reply:
Not currently. I think changing kubectl to add the context there may be most useful, as there may be other webhooks doing the rejecting.

Reply:
I think the template author should decide what type of information they want to return to end users either to guide them by providing the expected spec or some information to help remediate if it's identified during an audit.

Reply:
Is this different from what I'm suggesting? The template author has control over the messaging regarding the nature of the failure, the framework + enforcement point have control over the messaging of which resource fails.

Reply:
maybe we are saying the same thing but in different ways? 😄 

>  the framework + enforcement point have control over the messaging of which resource fails.

Since a constraint template can be evaluating multiple resources from a single request, it makes sense that the template author decides how to get the violating resource bubbled up all the way to the message users see.   For example, let's say an admission request has a deployment resource, but the violating resource could be one of the pods. So kubectl is going to fail on the deployment resource, but the template message should return that it was the pod that is causing the violation.

Reply:
I think we are saying different things in different ways.

In your example of a pod in a deployment, the pod template causes the deployment resource to fail. If it's failing because of some enumerated resource (e.g. container), then it makes sense to report which container as that is information about the structure of the resource itself, which is something no framework is expected to know. However, information about the identity of the deployment itself should be handled by kubectl, as it *is* expected to know what resources it is modifying.

TL;DR: for your deployment example, I agree. What I would disagree with is if people embedded the GVK/namespace/name of the deployment itself into the error.